File: src/main/java/org/luxons/sevenwonders/game/Lobby.java
package org.luxons.sevenwonders.game;

import java.util.ArrayList;
import java.util.List;

import org.luxons.sevenwonders.game.api.CustomizableSettings;
import org.luxons.sevenwonders.game.data.GameDefinition;

public class Lobby {

    private final long id;

    private final String name;

    private final Player owner;

    private final GameDefinition gameDefinition;

    private final List<Player> players;

    private CustomizableSettings settings;

    private State state = State.LOBBY;

    public Lobby(long id, String name, Player owner, GameDefinition gameDefinition) {
        this.id = id;
        this.name = name;
        this.owner = owner;
        this.gameDefinition = gameDefinition;
        this.players = new ArrayList<>(gameDefinition.getMinPlayers());
        this.settings = new CustomizableSettings();
        players.add(owner);
    }

    public long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public List<Player> getPlayers() {
        return players;
    }

    public CustomizableSettings getSettings() {
        return settings;
    }

    public void setSettings(CustomizableSettings settings) {
        this.settings = settings;
    }

    public synchronized void addPlayer(Player player) throws GameAlreadyStartedException, PlayerOverflowException {
        if (hasStarted()) {
            throw new GameAlreadyStartedException();
        }
        if (maxPlayersReached()) {
            throw new PlayerOverflowException();
        }
        if (playerNameAlreadyUsed(player.getDisplayName())) {
            throw new PlayerNameAlreadyUsedException(player.getDisplayName());
        }
        player.setIndex(players.size());
        players.add(player);
    }

    private boolean hasStarted() {
        return state != State.LOBBY;
    }

    private boolean maxPlayersReached() {
        return players.size() >= gameDefinition.getMaxPlayers();
    }

    private boolean playerNameAlreadyUsed(String name) {
        return players.stream().anyMatch(p -> p.getDisplayName().equals(name));
    }

    public synchronized Game startGame() throws PlayerUnderflowException {
        if (!hasEnoughPlayers()) {
            throw new PlayerUnderflowException();
        }
        state = State.PLAYING;
        return gameDefinition.initGame(id, settings, players);
    }

    private boolean hasEnoughPlayers() {
        return players.size() >= gameDefinition.getMinPlayers();
    }

    public void reorderPlayers(List<String> orderedUsernames) {
        List<Player> formerList = new ArrayList<>(players);
        players.clear();
        for (int i = 0; i < orderedUsernames.size(); i++) {
            Player player = getPlayer(formerList, orderedUsernames.get(i));
            players.add(player);
            player.setIndex(i);
        }
    }

    private static Player getPlayer(List<Player> players, String username) {
        return players.stream()
                      .filter(p -> p.getUsername().equals(username))
                      .findAny()
                      .orElseThrow(() -> new UnknownPlayerException(username));
    }

    public boolean isOwner(String username) {
        return owner.getUsername().equals(username);
    }

    public boolean containsUser(String username) {
        return players.stream().anyMatch(p -> p.getUsername().equals(username));
    }

    static class GameAlreadyStartedException extends IllegalStateException {
    }

    static class PlayerOverflowException extends IllegalStateException {
    }

    static class PlayerUnderflowException extends IllegalStateException {
    }

    static class PlayerNameAlreadyUsedException extends RuntimeException {
        PlayerNameAlreadyUsedException(String name) {
            super(name);
        }
    }

    static class UnknownPlayerException extends IllegalArgumentException {
        UnknownPlayerException(String username) {
            super(username);
        }
    }
}

File: sw-server/src/main/kotlin/org/luxons/sevenwonders/server/lobby/Lobby.kt
package org.luxons.sevenwonders.server.lobby

import org.luxons.sevenwonders.engine.Game
import org.luxons.sevenwonders.engine.data.GameDefinition
import org.luxons.sevenwonders.model.Settings
import org.luxons.sevenwonders.model.api.State
import org.luxons.sevenwonders.model.wonders.AssignedWonder
import org.luxons.sevenwonders.model.wonders.PreGameWonder
import org.luxons.sevenwonders.model.wonders.withRandomSide

class Lobby(
    val id: Long,
    val name: String,
    var owner: Player,
    private val gameDefinition: GameDefinition,
) {
    private val players: MutableList<Player> = ArrayList(gameDefinition.maxPlayers)

    val allWonders: List<PreGameWonder> = gameDefinition.allWonders

    private val assignedWonders: MutableList<AssignedWonder> = mutableListOf()

    var settings: Settings = Settings()

    var state = State.LOBBY
        private set

    init {
        addPlayer(owner)
    }

    fun getPlayers(): List<Player> = players

    fun getAssignedWonders(): List<AssignedWonder> = assignedWonders

    @Synchronized
    fun addPlayer(player: Player) {
        if (hasStarted()) {
            throw GameAlreadyStartedException(name)
        }
        if (maxPlayersReached()) {
            throw PlayerOverflowException(gameDefinition.maxPlayers)
        }
        if (playerNameAlreadyUsed(player.displayName)) {
            throw PlayerNameAlreadyUsedException(player.displayName, name)
        }
        player.join(this)
        players.add(player)
        assignedWonders.add(pickRandomWonder())
    }

    private fun pickRandomWonder(): AssignedWonder = allWonders.filter { !it.isAssigned() }.random().withRandomSide()

    private fun PreGameWonder.isAssigned() = name in assignedWonders.map { it.name }

    private fun hasStarted(): Boolean = state != State.LOBBY

    fun maxPlayersReached(): Boolean = players.size >= gameDefinition.maxPlayers

    private fun playerNameAlreadyUsed(name: String): Boolean = players.any { it.displayName == name }

    @Synchronized
    fun startGame(): Game {
        if (!hasEnoughPlayers()) {
            throw PlayerUnderflowException(gameDefinition.minPlayers)
        }
        state = State.PLAYING
        val game = gameDefinition.createGame(id, assignedWonders, settings)
        players.forEachIndexed { index, player -> player.join(game, index) }
        return game
    }

    fun hasEnoughPlayers(): Boolean = players.size >= gameDefinition.minPlayers

    @Synchronized
    fun reorderPlayers(orderedUsernames: List<String>) {
        val usernames = players.map { it.username }
        if (orderedUsernames.toSet() != usernames.toSet()) {
            throw PlayerListMismatchException(orderedUsernames)
        }
        val wondersMap = players.indices.associate { assignedWonders[it] to players[it].username }
        players.sortBy { orderedUsernames.indexOf(it.username) }
        assignedWonders.sortBy { orderedUsernames.indexOf(wondersMap[it]) }
    }

    @Synchronized
    fun reassignWonders(wonders: List<AssignedWonder>) {
        require(wonders.size == players.size)
        wonders.forEach {
            require(it.name in allWonders.map { w -> w.name })
        }
        assignedWonders.clear()
        assignedWonders.addAll(wonders)
    }

    @Synchronized
    fun isOwner(username: String?): Boolean = owner.username == username

    @Synchronized
    fun containsUser(username: String): Boolean = players.any { it.username == username }

    @Synchronized
    fun removePlayer(username: String): Player {
        val playerIndex = players.indexOfFirst { it.username == username }
        if (playerIndex < 0) {
            throw UnknownPlayerException(username)
        }
        assignedWonders.removeAt(playerIndex)
        val player = players.removeAt(playerIndex)
        player.leave()

        if (player == owner && players.isNotEmpty()) {
            owner = players[0]
        }
        return player
    }

    fun setEndOfGame() {
        state = State.FINISHED
    }

    internal class GameAlreadyStartedException(name: String) : IllegalStateException("Game '$name' has already started")

    internal class PlayerOverflowException(max: Int) : IllegalStateException("Maximum $max players allowed")

    internal class PlayerUnderflowException(min: Int) :
        IllegalStateException("Minimum $min players required to start a game")

    internal class PlayerNameAlreadyUsedException(displayName: String, gameName: String) :
        IllegalArgumentException("Name '$displayName' is already used by a player in game '$gameName'")

    internal class UnknownPlayerException(username: String) : IllegalArgumentException("Unknown player '$username'")

    internal class PlayerListMismatchException(usernames: List<String>) :
        IllegalArgumentException("Newly ordered usernames $usernames don't match the current player list")
}
